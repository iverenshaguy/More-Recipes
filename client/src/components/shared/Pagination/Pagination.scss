@import '../../../assets/scss/modules';

.pagination {
  .page-item {
    &.active {
      .page-link {
        background-color: $green;
        border-color: $green;
        color: $white;
        z-index: 2;
      }
    }
  }
}
