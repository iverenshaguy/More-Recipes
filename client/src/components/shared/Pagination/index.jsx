import React, { Component } from 'react';
import { Pagination as PaginationWrapper, PaginationItem, PaginationLink } from 'reactstrap';
import PropTypes from 'prop-types';
import { multiRecipePropTypes } from '../../../helpers/proptypes';
import './Pagination.scss';

/**
 * @exports
 * @class Pagination
 * @extends Component
 * @returns {component} Pagination
 */
class Pagination extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.any).isRequired,
    onPageChange: PropTypes.func.isRequired,
    ...multiRecipePropTypes.metadata
  };

  /**
   * @constructor
   * @memberof Pagination
   * @param {object} props
   * @returns {nothing} - Returns nothing
   */
  constructor(props) {
    super(props);

    this.state = {
      pager: {}
    };
  }

  /**
   * @memberof Pagination
   * @return {nothing} Returns nothing
   */
  componentWillMount() {
    const { metadata } = this.props;
    const {
      totalCount, page, itemsPerPage, firstPage, lastPage, pages
    } = metadata;

    if (this.props.items && this.props.items.length) {
      this.setState({
        pager: {
          initialPage: firstPage,
          currentPage: page,
          totalPages: pages.length,
          pages,
          itemsPerPage,
          totalCount,
          lastPage
        }
      });
    }
  }

  /**
   * @memberof Pagination
   * @param {number} page
   * @return {nothing} Returns nothing
   */
  setPage(page) {
    // call change page function in parent component
    this.props.onPageChange(page);
  }

  /**
   * @memberof Pagination
   * @returns {component} Pagination
   */
  render() {
    const { pager } = this.state;

    if (!pager.pages || pager.pages.length <= 1) {
      // don't display pager if there is only 1 page
      return null;
    }

    return (
      <div className="mt-5">
        <nav aria-label="Page navigation">
          <PaginationWrapper className="justify-content-center">
            <PaginationItem className={`${pager.currentPage === 1 ? 'disabled' : ''}`}>
              <PaginationLink onClick={() => this.setPage(1)}>
                First
              </PaginationLink>
            </PaginationItem>
            <PaginationItem className={`${pager.currentPage === 1 ? 'disabled' : ''}`}>
              <PaginationLink onClick={() => this.setPage(pager.currentPage - 1)}>
                <span aria-hidden="true">&laquo;</span>
                <span className="sr-only">Previous</span>
              </PaginationLink>
            </PaginationItem>
            {pager.pages.map(page => (
              <PaginationItem
                key={page}
                className={`${pager.currentPage === page ? 'active' : ''}`}
              >
                <PaginationLink onClick={() => this.setPage(page)}>
                  {page}
                </PaginationLink>
              </PaginationItem>
            ))}
            <PaginationItem className={`${pager.currentPage === pager.totalPages ? 'disabled' : ''}`}>
              <PaginationLink onClick={() => this.setPage(pager.currentPage + 1)}>
                <span aria-hidden="true">&raquo;</span>
                <span className="sr-only">Next</span>
              </PaginationLink>
            </PaginationItem>
            <PaginationItem className={`${pager.currentPage === pager.totalPages ? 'disabled' : ''}`}>
              <PaginationLink onClick={() => this.setPage(pager.totalPages)}>
                Last
              </PaginationLink>
            </PaginationItem>
          </PaginationWrapper>
        </nav>
      </div>
    );
  }
}

export default Pagination;
