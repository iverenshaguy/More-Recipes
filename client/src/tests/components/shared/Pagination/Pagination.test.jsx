import React from 'react';
import Pagination from '../../../../components/shared/Pagination';

const props = {
  items: [{ id: 1, recipeName: 'Rice' }, { id: 5, recipeName: 'Beans' }],
  onPageChange: jest.fn(),
  metadata: {
    firstPage: 1,
    lastPage: 3,
    page: 2,
    itemsPerPage: 5,
    pages: [1, 2, 3],
    totalCount: 13,
  }
};

const currentPageProps = {
  items: [{ id: 1, recipeName: 'Rice' }, { id: 5, recipeName: 'Beans' }],
  onPageChange: jest.fn(),
  metadata: {
    firstPage: 1,
    lastPage: 3,
    page: 1,
    itemsPerPage: 5,
    pages: [1, 2, 3],
    totalCount: 13,
  }
};

const lastPageProps = {
  items: [{ id: 1, recipeName: 'Rice' }, { id: 5, recipeName: 'Beans' }],
  onPageChange: jest.fn(),
  metadata: {
    firstPage: 1,
    lastPage: 3,
    page: 3,
    itemsPerPage: 5,
    pages: [1, 2, 3],
    totalCount: 13,
  }
};

const metadata = {
  firstPage: 1,
  lastPage: 1,
  page: 1,
  itemsPerPage: 2,
  pages: [1],
  totalCount: 2,
};

describe('Pagination', () => {
  it('renders correctly', () => {
    const mountedWrapper = mount(<Pagination {...props} />);
    expect(toJson(mountedWrapper)).toMatchSnapshot();
    mountedWrapper.unmount();
  });

  it('renders correctly when there is only one page', () => {
    const mountedWrapper = mount(<Pagination {...props} metadata={metadata} />);
    expect(toJson(mountedWrapper)).toMatchSnapshot();
    mountedWrapper.unmount();
  });

  it('disables first and previous links when current page == firstpage', () => {
    const mountedWrapper = mount(<Pagination {...currentPageProps} />);

    expect(toJson(mountedWrapper)).toMatchSnapshot();
    expect(mountedWrapper.find('li.page-item').at(0).hasClass('disabled')).toBeTruthy();
    expect(mountedWrapper.find('li.page-item').at(1).hasClass('disabled')).toBeTruthy();
    expect(mountedWrapper.find('li.page-item').at(5).hasClass('disabled')).toBeFalsy();
    expect(mountedWrapper.find('li.page-item').at(6).hasClass('disabled')).toBeFalsy();
    mountedWrapper.unmount();
  });

  it('goes to next page when next button is clicked', () => {
    const mountedWrapper = mount(<Pagination {...currentPageProps} />);

    const setPageSpy = jest.spyOn(mountedWrapper.instance(), 'setPage');

    mountedWrapper.find('a.page-link').at(5).simulate('click');

    expect(setPageSpy).toHaveBeenCalledWith(2);
    mountedWrapper.unmount();
  });

  it('goes to last page when last button is clicked', () => {
    const mountedWrapper = mount(<Pagination {...currentPageProps} />);

    const setPageSpy = jest.spyOn(mountedWrapper.instance(), 'setPage');

    mountedWrapper.find('a.page-link').at(6).simulate('click');

    expect(setPageSpy).toHaveBeenCalledWith(3);
    mountedWrapper.unmount();
  });

  it('disables first and previous links when current page == lastpage', () => {
    const mountedWrapper = mount(<Pagination {...lastPageProps} />);

    expect(toJson(mountedWrapper)).toMatchSnapshot();
    expect(mountedWrapper.find('li.page-item').at(5).hasClass('disabled')).toBeTruthy();
    expect(mountedWrapper.find('li.page-item').at(6).hasClass('disabled')).toBeTruthy();
    expect(mountedWrapper.find('li.page-item').at(0).hasClass('disabled')).toBeFalsy();
    expect(mountedWrapper.find('li.page-item').at(1).hasClass('disabled')).toBeFalsy();
    mountedWrapper.unmount();
  });

  it('goes to previous page when previous button is clicked', () => {
    const mountedWrapper = mount(<Pagination {...lastPageProps} />);

    const setPageSpy = jest.spyOn(mountedWrapper.instance(), 'setPage');

    mountedWrapper.find('a.page-link').at(1).simulate('click');

    expect(setPageSpy).toHaveBeenCalledWith(2);
    mountedWrapper.unmount();
  });

  it('goes to first page when first button is clicked', () => {
    const mountedWrapper = mount(<Pagination {...lastPageProps} />);

    const setPageSpy = jest.spyOn(mountedWrapper.instance(), 'setPage');

    mountedWrapper.find('a.page-link').at(0).simulate('click');

    expect(setPageSpy).toHaveBeenCalledWith(1);
    mountedWrapper.unmount();
  });

  it('goes to correct page when clicked: page 1', () => {
    const mountedWrapper = mount(<Pagination {...props} />);

    const setPageSpy = jest.spyOn(mountedWrapper.instance(), 'setPage');

    mountedWrapper.find('a.page-link').at(3).simulate('click');
    mountedWrapper.find('a.page-link').at(2).simulate('click');

    expect(setPageSpy).toHaveBeenCalledWith(1);
    mountedWrapper.unmount();
  });

  it('goes to correct page when clicked: page 2', () => {
    const mountedWrapper = mount(<Pagination {...props} />);

    const setPageSpy = jest.spyOn(mountedWrapper.instance(), 'setPage');

    mountedWrapper.find('a.page-link').at(3).simulate('click');

    expect(setPageSpy).toHaveBeenCalledWith(2);
    mountedWrapper.unmount();
  });

  it('goes to correct page when clicked: page 3', () => {
    const mountedWrapper = mount(<Pagination {...props} />);

    const setPageSpy = jest.spyOn(mountedWrapper.instance(), 'setPage');

    mountedWrapper.find('a.page-link').at(4).simulate('click');

    expect(setPageSpy).toHaveBeenCalledWith(3);
    mountedWrapper.unmount();
  });
});
