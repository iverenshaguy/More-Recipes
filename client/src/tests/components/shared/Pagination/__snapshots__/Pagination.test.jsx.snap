// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination disables first and previous links when current page == firstpage 1`] = `
<Pagination
  items={
    Array [
      Object {
        "id": 1,
        "recipeName": "Rice",
      },
      Object {
        "id": 5,
        "recipeName": "Beans",
      },
    ]
  }
  metadata={
    Object {
      "firstPage": 1,
      "itemsPerPage": 5,
      "lastPage": 3,
      "page": 1,
      "pages": Array [
        1,
        2,
        3,
      ],
      "totalCount": 13,
    }
  }
  onPageChange={[Function]}
>
  <div
    className="mt-5"
  >
    <nav
      aria-label="Page navigation"
    >
      <Pagination
        className="justify-content-center"
        tag="ul"
      >
        <ul
          className="justify-content-center pagination"
        >
          <PaginationItem
            className="disabled"
            tag="li"
          >
            <li
              className="disabled page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  First
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className="disabled"
            tag="li"
          >
            <li
              className="disabled page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden="true"
                  >
                    «
                  </span>
                  <span
                    className="sr-only"
                  >
                    Previous
                  </span>
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className="active"
            key="1"
            tag="li"
          >
            <li
              className="active page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  1
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            key="2"
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  2
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            key="3"
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  3
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden="true"
                  >
                    »
                  </span>
                  <span
                    className="sr-only"
                  >
                    Next
                  </span>
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  Last
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
        </ul>
      </Pagination>
    </nav>
  </div>
</Pagination>
`;

exports[`Pagination disables first and previous links when current page == lastpage 1`] = `
<Pagination
  items={
    Array [
      Object {
        "id": 1,
        "recipeName": "Rice",
      },
      Object {
        "id": 5,
        "recipeName": "Beans",
      },
    ]
  }
  metadata={
    Object {
      "firstPage": 1,
      "itemsPerPage": 5,
      "lastPage": 3,
      "page": 3,
      "pages": Array [
        1,
        2,
        3,
      ],
      "totalCount": 13,
    }
  }
  onPageChange={[Function]}
>
  <div
    className="mt-5"
  >
    <nav
      aria-label="Page navigation"
    >
      <Pagination
        className="justify-content-center"
        tag="ul"
      >
        <ul
          className="justify-content-center pagination"
        >
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  First
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden="true"
                  >
                    «
                  </span>
                  <span
                    className="sr-only"
                  >
                    Previous
                  </span>
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            key="1"
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  1
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            key="2"
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  2
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className="active"
            key="3"
            tag="li"
          >
            <li
              className="active page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  3
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className="disabled"
            tag="li"
          >
            <li
              className="disabled page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden="true"
                  >
                    »
                  </span>
                  <span
                    className="sr-only"
                  >
                    Next
                  </span>
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className="disabled"
            tag="li"
          >
            <li
              className="disabled page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  Last
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
        </ul>
      </Pagination>
    </nav>
  </div>
</Pagination>
`;

exports[`Pagination renders correctly 1`] = `
<Pagination
  items={
    Array [
      Object {
        "id": 1,
        "recipeName": "Rice",
      },
      Object {
        "id": 5,
        "recipeName": "Beans",
      },
    ]
  }
  metadata={
    Object {
      "firstPage": 1,
      "itemsPerPage": 5,
      "lastPage": 3,
      "page": 2,
      "pages": Array [
        1,
        2,
        3,
      ],
      "totalCount": 13,
    }
  }
  onPageChange={[Function]}
>
  <div
    className="mt-5"
  >
    <nav
      aria-label="Page navigation"
    >
      <Pagination
        className="justify-content-center"
        tag="ul"
      >
        <ul
          className="justify-content-center pagination"
        >
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  First
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden="true"
                  >
                    «
                  </span>
                  <span
                    className="sr-only"
                  >
                    Previous
                  </span>
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            key="1"
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  1
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className="active"
            key="2"
            tag="li"
          >
            <li
              className="active page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  2
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            key="3"
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  3
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  <span
                    aria-hidden="true"
                  >
                    »
                  </span>
                  <span
                    className="sr-only"
                  >
                    Next
                  </span>
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
          <PaginationItem
            className=""
            tag="li"
          >
            <li
              className="page-item"
            >
              <PaginationLink
                onClick={[Function]}
                tag="a"
              >
                <a
                  className="page-link"
                  onClick={[Function]}
                >
                  Last
                </a>
              </PaginationLink>
            </li>
          </PaginationItem>
        </ul>
      </Pagination>
    </nav>
  </div>
</Pagination>
`;

exports[`Pagination renders correctly when there is only one page 1`] = `
<Pagination
  items={
    Array [
      Object {
        "id": 1,
        "recipeName": "Rice",
      },
      Object {
        "id": 5,
        "recipeName": "Beans",
      },
    ]
  }
  metadata={
    Object {
      "firstPage": 1,
      "itemsPerPage": 2,
      "lastPage": 1,
      "page": 1,
      "pages": Array [
        1,
      ],
      "totalCount": 2,
    }
  }
  onPageChange={[Function]}
/>
`;
